libcudacxx/include/cuda/std/__random/philox_engine.h

```diff
@@ -0,0 +1,575 @@
+    const _CharT __fill                       = __os.fill();
+
+    __os.flags(ios_base::dec | ios_base::fixed | ios_base::left);
+    __os.fill(__os.widen(' '));
+
+    // output counter array (__x_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __os << __e.__x_[__i];
+      if (__i < word_count - 1)
+      {
+        __os << __os.widen(' ');
+      }
+    }
+    __os << __os.widen(' ');
+
+    // output key array (__k_)
+    for (::cuda::std::size_t __i = 0; __i < word_count / 2; ++__i)
+    {
+      __os << __e.__k_[__i];
+      if (__i < word_count / 2 - 1)
+      {
+        __os << __os.widen(' ');
+      }
+    }
+    __os << __os.widen(' ');
+
+    // output output buffer (__y_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __os << __e.__y_[__i];
+      if (__i < word_count - 1)
+      {
+        __os << __os.widen(' ');
+      }
+    }
+    __os << __os.widen(' ');
+
+    // output current position
+    __os << __e.__j_;
+
+    // restore flags & fill character
+    __os.flags(__flags);
+    __os.fill(__fill);
+
+    return __os;
+  }
+
+  //! This function streams a philox_engine in from a std::basic_istream.
+  //! @param is The basic_istream to stream from.
+  //! @param e The philox_engine to stream in.
+  //! @return is
+  template <typename _CharT, typename _Traits>
+  _CCCL_API friend ::std::basic_istream<_CharT, _Traits>&
+  operator>>(::std::basic_istream<_CharT, _Traits>& __is, philox_engine& __e)
+  {
+    using istream_type = ::std::basic_istream<_CharT, _Traits>;
+    using ios_base     = typename istream_type::ios_base;
+
+    // save old flags
+    const typename ios_base::fmtflags __flags = __is.flags();
+
+    __is.flags(ios_base::dec);
+
+    // input counter array (__x_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __is >> __e.__x_[__i];
+    }
+
+    // input key array (__k_)
+    for (::cuda::std::size_t __i = 0; __i < word_count / 2; ++__i)
+    {
+      __is >> __e.__k_[__i];
+    }
+
+    // input output buffer (__y_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __is >> __e.__y_[__i];
+    }
+
+    // input current position
+    __is >> __e.__j_;
+
+    // restore flags
+    __is.flags(__flags);
+
+    return __is;
+  }
+#endif // !_CCCL_COMPILER(NVRTC)
+
+private:
+  _CCCL_API constexpr void __increment_counter() noexcept
+  {
+    // Increment the big integer __x_ by 1, handling overflow.
+    _CCCL_PRAGMA_UNROLL_FULL()
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __x_[__i] = (__x_[__i] + 1) & max();
+      if (__x_[__i] != 0)
+      {
+        break;
+      }
+    }
+  }
+
+  [[nodiscard]] _CCCL_API constexpr auto __mulhilo_fallback(result_type __a, result_type __b) const noexcept
+  {
+    // Generic slow implementation
+    constexpr result_type __w_half  = word_size / 2;
+    constexpr result_type __lo_mask = (((result_type) 1) << __w_half) - 1;
```

Comment:
We should avoid `C` cases
```suggestion
    constexpr result_type __lo_mask = (result_type{1} << __w_half) - 1;
```